src/test/kotlin/com/zepben/evolve/streaming/get/UpdateNetworkStateClientTest.kt

```diff
@@ -0,0 +1,145 @@
+                partialFailure = true
+                addAllFailed(listOf(
+                    PBStateEventFailure.newBuilder().apply {
+                        eventId = "event1"
+                        unknownMrid = PBStateEventUnknownMrid.newBuilder().build()
+                    }.build(),
+                    PBStateEventFailure.newBuilder().apply {
+                        eventId = "event2"
+                        duplicateMrid = PBStateEventDuplicateMrid.newBuilder().build()
+                    }.build(),
+                    PBStateEventFailure.newBuilder().apply {
+                        eventId = "event3"
+                        invalidMrid = PBStateEventInvalidMrid.newBuilder().build()
+                    }.build(),
+                    PBStateEventFailure.newBuilder().apply {
+                        eventId = "event4"
+                        unsupportedPhasing = PBStateEventUnsupportedPhasing.newBuilder().build()
+                    }.build()))
+            }.build() }.build(),
+            null
+        ))
+        service.onSetCurrentStates = spy { request, response ->
+            request.eventList.let {
+                assertThat(it.map { it.eventId }, anyOf(switchStateEvents.map { hasItem(it.eventId) }))
+                assertThat(it.map { it.timestamp }, anyOf(switchStateEvents.map { hasItem(it.timestamp.toTimestamp()) }))
+                assertThat(it.map { it.switch.mrid }, anyOf(switchStateEvents.map { hasItem(it.mRID) }))
+                assertThat(it.map { it.switch.action.name }, anyOf(switchStateEvents.map { hasItem(it.action.name) }))
+                assertThat(it.map { it.switch.phases.name }, anyOf(switchStateEvents.map { hasItem(it.phases.name) }))
+            }
+            response.onNext(responses.removeFirst())
+        }
+
+        act()
+    }
+
+    private fun assertBatchedCurrentStatesStatus(currentStatesStatus: List<SetCurrentStatesStatus>){
+        assertThat(currentStatesStatus.size, equalTo(3))
+        assertThat(currentStatesStatus[0], instanceOf(BatchSuccessful::class.java))
+        (currentStatesStatus[1] as ProcessingPaused).let {
+            assertThat(it.since, equalTo(timestampOf1Second.toLocalDateTime()))
+        }
+        (currentStatesStatus[2] as BatchFailure).let {
+            assertThat(it.partialFailure, equalTo(true))
+            assertThat(it.failures.size, equalTo(4))
+            it.failures.assertFailure(0, StateEventUnknownMrid::class.java)
+            it.failures.assertFailure(1, StateEventDuplicateMrid::class.java)
+            it.failures.assertFailure(2, StateEventInvalidMrid::class.java)
+            it.failures.assertFailure(3, StateEventUnsupportedPhasing::class.java)
+        }
+    }
+
+    private fun List<StateEventFailure>.assertFailure(index: Int, clazz: Class<*>){
+        this[index].let {
+            assertThat(it.eventId, equalTo("event${index + 1}"))
+            assertThat(it, instanceOf(clazz))
+        }
+    }
```

Comment:
suggestion: You could make this work directly on `BatchFailure` rather than the `failures` list, which would reduce the amount of repeated code in the test and replace the `this[index]` call here with `failures[index]`
```suggestion
    private fun BatchFailure.assertFailure(index: Int, clazz: Class<*>) {
        failures[index].also {
            assertThat(it.eventId, equalTo("event${index + 1}"))
            assertThat(it, instanceOf(clazz))
        }
    }
```
with usage:
```kotlin
it.assertFailure(0, StateEventUnknownMrid::class.java)
it.assertFailure(1, StateEventDuplicateMrid::class.java)
it.assertFailure(2, StateEventInvalidMrid::class.java)
it.assertFailure(3, StateEventUnsupportedPhasing::class.java)
```

You take this one step further and use an inline reified function:
```suggestion
    private inline fun <reified T> BatchFailure.assertFailure(index: Int) {
        failures[index].also {
            assertThat(it.eventId, equalTo("event${index + 1}"))
            assertThat(it, instanceOf(T::class.java))
        }
    }
```
with usage:
```kotlin
it.assertFailure<StateEventUnknownMrid>(0)
it.assertFailure<StateEventDuplicateMrid>(1)
it.assertFailure<StateEventInvalidMrid>(2)
it.assertFailure<StateEventUnsupportedPhasing>(3)
```

alternatively, you could just remove this function and test the failures collection directly:
```kotlin
assertThat(it.partialFailure, equalTo(true))
assertThat(
    it.failures.map { f -> f::class },
    contains(
        StateEventUnknownMrid::class,
        StateEventDuplicateMrid::class,
        StateEventInvalidMrid::class,
        StateEventUnsupportedPhasing::class
    )
)
it.failures.forEachIndexed { index, f ->
    assertThat(f.eventId, equalTo("event${index + 1}"))
}
```
noting that `contains` check the collection has the exact size and order, so no need to test the size separately.